go-controller/pkg/ovn/base_network_controller_pods.go

```diff
@@ -574,12 +574,14 @@ func (bnc *BaseNetworkController) addLogicalPortToNetwork(pod *kapi.Pod, nadName
 		// IP/MAC from the annotation.
 		lsp.DynamicAddresses = nil
 
-		// ensure we have reserved the IPs in the annotation
-		if err = bnc.lsManager.AllocateIPs(switchName, podIfAddrs); err != nil && err != ipallocator.ErrAllocated {
-			return nil, nil, nil, false, fmt.Errorf("unable to ensure IPs allocated for already annotated pod: %s, IPs: %s, error: %v",
-				podDesc, util.JoinIPNetIPs(podIfAddrs, " "), err)
-		} else {
-			needsIP = false
+		if bnc.doesNetworkRequireIPAM() {
+			// ensure we have reserved the IPs in the annotation
+			if err = bnc.lsManager.AllocateIPs(switchName, podIfAddrs); err != nil && err != ipallocator.ErrAllocated {
```

Comment:
`AllocateIPs` could fail if we don't have IPAM and so we have a MAC annotated but no IPs annotated?

Reply:
Should we handle this scenario so that it doesn't fail? It could happen that we process a pod update after we have annotated the mac on a previous pod event ?

Reply:
Also as we talked, we would need to consider the case where a user directly annotated a MAC that could be potentially used by a different pod.

Reply:
> Should we handle this scenario so that it doesn't fail? It could happen that we process a pod update after we have annotated the mac on a previous pod event ?

I don't follow. Is this a legit use case ? Let's say you have a pod running, the user updates its IP address by updating the annotation. Is this allowed ? Do we expect the pod's IP to change ? Reading the code ... it seems it can happen.

And what do you mean with "handle this scenario so that it doesn't fail?" ? Can you elaborate on what's the scenario ? A user providing an IPs / MACs via the annotation ? A user **updating** an existing pod's annotation ? 

I feel I'm missing something, but can't quite grasp what. 

Lastly, for what it's worth, now that we have an API to allow the user to request IPs / MACs - the network-selection-elements - it feels weird to allow specifying the same via annotations. It's confusing, and at best raising questions: which takes precedence ? Annotation vs network-selection-elements .

Reply:
I don't understand the original comment "AllocateIPs could fail if we don't have IPAM, we have a MAC annotated but no IPs annotated?"

IMO this cannot happen. We wont try to allocate IPs until IPAM is present. We ensure this earlier in the function by:  bnc.waitForNodeLogicalSwitchSubnetsInCache

Reply:
> Lastly, for what it's worth, now that we have an API to allow the user to request IPs / MACs - the network-selection-elements - it feels weird to allow specifying the same via annotations. It's confusing, and at best raising questions: which takes precedence ? Annotation vs network-selection-elements.

We don't allow the **user** to specify  IPs / MACs through annotations. That's there set by ovn-k for it's own reference.


> I don't follow. Is this a legit use case ? Let's say you have a pod running, the user updates its IP address by updating the annotation. Is this allowed ? Do we expect the pod's IP to change ? Reading the code ... it seems it can happen.
> 
> And what do you mean with "handle this scenario so that it doesn't fail?" ? Can you elaborate on what's the scenario ? A user providing an IPs / MACs via the annotation ? A user **updating** an existing pod's annotation ?
> 
> I feel I'm missing something, but can't quite grasp what.
> 

It's not a use case because it is nothing the user should trigger.

Let's say that we process a create pod event. We allocate and annotate a MAC to the pod. But a few lines of code later, something goes wrong and we queue a retry. When we do the retry, the pod already has annotated the MAC that we previously allocated. We need to be able to handle the retry of the pod creation idem-potently. You will find yourself with a pod that already has a MAC annotated but with other stuff pending to be done for it to be attached.

Reply:
Okay I checked the code again and I don't see that we hit `AllocateIPs` in the scenario I explained above (for layer 2 network without IPAM). I don't know why I saw that before.

Reply:
What I do see now though is that for the scenario above:

needsIP will be true, needsNewMacOrIPAllocation will be true,  assignPodAddresses will be called where a new MAC address will be allocated replacing the one we previously allocated.

Unsure if this is a problem down the line but should be easy enough to prevent.

Reply:
> We don't allow the **user** to specify IPs / MACs through annotations. That's there set by ovn-k for it's own reference.
> 
> Understand `allow` as: we don't support it but we should protect against to the best of our ability because we can't prevent it.

I should clarify there is very little we can do to prevent someone from annotating a MAC in the first place and telling that apart from the case where the MAC has been annotated by OVN-K.

That's why it could be useful to keep track of the MACs that OVN-K has allocated and fail the annotated MAC we get has been previously assigned to a different pod.